File: dashboard/static/portfolio-opt.js
/* ── 组合优化模块 ── */

const PortfolioOpt = {
    _methods: [],

    async init() {
        if (this._methods.length) return;
        try {
            const data = await App.fetchJSON('/api/portfolio-opt/methods');
            this._methods = data.methods || [];
        } catch (e) {
            console.error('加载优化方法失败:', e);
        }
    },

    async optimize() {
        await this.init();
        const codesStr = document.getElementById('portopt-codes')?.value || '';
        const codes = codesStr.split(/[,，\s]+/).filter(Boolean);
        const method = document.getElementById('portopt-method')?.value || 'max_sharpe';
        const startDate = document.getElementById('portopt-start')?.value || '2024-01-01';
        const endDate = document.getElementById('portopt-end')?.value || '2024-12-31';

        if (codes.length < 2) { App.toast('至少需要2只股票', 'error'); return; }

        const btn = App._getLegacyActionButton ? App._getLegacyActionButton('portopt-optimize') : null;
        if (btn) { btn.disabled = true; btn.textContent = '优化中...'; }

        try {
            const result = await App.postJSON('/api/portfolio-opt/optimize', {
                codes,
                method,
                start_date: startDate,
                end_date: endDate,
            });

            if (!result.success) { App.toast(result.error || '优化失败', 'error'); return; }

            // 更新指标
            document.getElementById('portopt-return').textContent = (result.expected_return * 100).toFixed(2) + '%';
            document.getElementById('portopt-vol').textContent = (result.expected_volatility * 100).toFixed(2) + '%';
            document.getElementById('portopt-sharpe').textContent = result.sharpe_ratio?.toFixed(4) || '--';

            // 权重饼图
            this._renderPie(result.weights || {});

            // 权重表格
            this._renderWeightTable(result.weights || {});

            App.toast('组合优化完成', 'success');
        } catch (e) {
            App.toast('组合优化失败: ' + e.message, 'error');
        } finally {
            if (btn) { btn.disabled = false; btn.textContent = '执行优化'; }
        }
    },

    _renderPie(weights) {
        const canvas = document.getElementById('portopt-pie-chart');
        if (!canvas) return;
        const ctx = canvas.getContext('2d');
        if (canvas._chart) canvas._chart.destroy();

        const entries = Object.entries(weights).filter(([, v]) => v > 0.001);
        const labels = entries.map(([k]) => k);
        const values = entries.map(([, v]) => (v * 100).toFixed(1));
        const palette = [
            'rgba(99,102,241,0.8)', 'rgba(16,185,129,0.8)', 'rgba(245,158,11,0.8)',
            'rgba(239,68,68,0.8)', 'rgba(139,92,246,0.8)', 'rgba(6,182,212,0.8)',
            'rgba(236,72,153,0.8)', 'rgba(132,204,22,0.8)', 'rgba(251,146,60,0.8)',
            'rgba(168,85,247,0.8)', 'rgba(20,184,166,0.8)', 'rgba(244,63,94,0.8)',
        ];

        canvas._chart = new Chart(ctx, {
            type: 'doughnut',
            data: {
                labels,
                datasets: [{
                    data: values,
                    backgroundColor: palette.slice(0, labels.length),
                    borderWidth: 2,
                    borderColor: 'rgba(15,23,42,0.8)',
                }],
            },
            options: {
                responsive: true,
                maintainAspectRatio: false,
                plugins: {
                    legend: { position: 'right', labels: { color: '#94a3b8', font: { size: 12 } } },
                    tooltip: { callbacks: { label: c => `${c.label}: ${c.parsed}%` } },
                },
            },
        });
    },

    _renderWeightTable(weights) {
        const table = document.getElementById('portopt-weight-table');
        if (!table) return;
        const entries = Object.entries(weights).filter(([, v]) => v > 0.001).sort((a, b) => b[1] - a[1]);
        let html = '<thead><tr><th>股票</th><th>权重</th></tr></thead><tbody>';
        entries.forEach(([code, w]) => {
            html += `<tr><td>${App.escapeHTML(code)}</td><td style="text-align:right">${(w * 100).toFixed(2)}%</td></tr>`;
        });
        html += '</tbody>';
        table.innerHTML = html;
    },
};

File: dashboard/routers/portfolio_opt.py
"""组合优化 API"""
from fastapi import APIRouter, HTTPException
from pydantic import BaseModel

from engine.portfolio_optimizer import METHODS, optimize_portfolio

router = APIRouter()


class PortfolioOptRequest(BaseModel):
    codes: list[str]
    start_date: str = "2024-01-01"
    end_date: str = "2024-12-31"
    method: str = "max_sharpe"  # equal_weight | risk_parity | min_variance | max_sharpe
    risk_free: float = 0.03


@router.get("/methods")
async def list_methods():
    """获取可用优化方法"""
    return {
        "methods": [
            {"name": "equal_weight", "label": "等权组合", "description": "每只股票等权重配置"},
            {"name": "risk_parity", "label": "风险平价", "description": "权重与波动率成反比"},
            {"name": "min_variance", "label": "最小方差", "description": "最小化组合方差"},
            {"name": "max_sharpe", "label": "最大夏普", "description": "最大化夏普比率"},
        ]
    }


@router.post("/optimize")
async def optimize(req: PortfolioOptRequest):
    """执行组合优化"""
    if len(req.codes) < 2:
        raise HTTPException(400, "至少需要2只股票")
    if len(req.codes) > 20:
        raise HTTPException(400, "单次最多优化20只股票")
    if req.method not in METHODS:
        raise HTTPException(400, f"不支持的优化方法: {req.method}")

    try:
        result = optimize_portfolio(
            codes=req.codes,
            start_date=req.start_date,
            end_date=req.end_date,
            method=req.method,
            risk_free=req.risk_free,
        )
        return {
            "success": True,
            "weights": result.weights,
            "expected_return": result.expected_return,
            "expected_volatility": result.expected_volatility,
            "sharpe_ratio": result.sharpe_ratio,
            "method": result.method,
        }
    except Exception as e:
        raise HTTPException(500, f"组合优化失败: {e}")
